### PR TITLE
[feature] driver/netio: use netio via json-api

### PR DESCRIPTION
**Pull request by ActionHOSchT**
**Description**

Following the contribution-guidelines an additional functionality using the netio-json-api was added to simplify netio handling. To get all actions working powerprotocol.py:NetworkPowerDriver probably needs a refactor.

Tested manually with a Netio-4-Device and in productive-env used with netio-8.
grafik
grafik 

Signed-off-by: Raffael Krakau <r.krakau90@gmail.com>

**Checklist**
- [x] Documentation for the feature
- [ ] Tests for the feature 

- [x] The arguments and description in doc/configuration.rst have been updated

- [x] PR has been tested
<!---

### CODE REVIEW

File: labgrid/driver/power/netio_json.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+Set model in YML-CFG to 'netio_json'.
+"""
+__author__ = "Eugen Wiens, Christian Happ, Raffael Krakau"
+__copyright__ = "Copyright 2023, JUMO GmbH & Co. KG"
+__email__ = "Eugen.Wiens@jumo.net, Christian.Happ@jumo.net"
+__date__ = "2023-07-05"
+__status__ = "Production"
+
+import json
+
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py:15]: Die Leerzeile könnte von mir aus weg

File: labgrid/driver/power/netio_json.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    Netio-Class to switch sockets via HTTP-JSON-API
+    """
+    __host: str
+    __port: int
+    __username: str
+    __password: str
+    __telnetConnection = None
+
+    def __init__(self, host, port=80, username=None, password=None):
+        """
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py]: this comment should you rework or remove, it gives not that much informations

File: labgrid/driver/power/netio_json.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                if outputState['ID'] == socketID:
+                    state = outputState
+                    break
+        else:
+            raise Exception(f"Cannot SET the power state for socket number: {socketID} for action: {action}."
+                            f" Error code: {response.text}")
+
+        return state
+
+    def get_state(self, socketID):
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py:62]: Funktioniert das mit dem get nun? Liegt es an diesem Treiber oder geht es mit dem `netio` power treiber auch nicht?

File: labgrid/driver/power/netio_json.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                    state = outputState
+                    break
+        else:
+            raise Exception(f"Cannot get the power state for socket {socketID}. Error code: {response.text}")
+
+        return state
+
+    def convert_socket_id(self, socketID) -> int:
+        """
+        cast socketID to int
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py]:
```suggestion
        Cast socketID to int
```

File: labgrid/driver/power/netio_json.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        try:
+            socketID = int(socketID)
+        except ValueError as e:
+            raise Exception(f"socketID \"{socketID}\" could not be converted to an integer: {e}!")
+
+        return socketID
+
+    def create_request_url(self) -> str:
+        """
+        Build URL from params set in __init__()
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py]:
```suggestion
        Generate request URL from given params
```

File: labgrid/driver/power/netio_json.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        if self.__port:
+            requestUrl += f':{self.__port}'
+
+        requestUrl += '/netio.json'
+
+        return requestUrl
+
+    def get_request_json_data(self, socketID, action) -> str:
+        data = f'{{"Outputs":[{{"ID":{socketID},"Action":{self.convert_action(action)}}}]}}'
+
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py]: remove the empty line, please

File: labgrid/driver/power/netio_json.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    def create_request_url(self) -> str:
+        """
+        Build URL from params set in __init__()
+        """
+        requestUrl = 'http://'
+
+        if self.__username and self.__password:
+            requestUrl += f'{self.__username}:{self.__password}@'
+
+        requestUrl += self.__host
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py]: The if statement to concat the port seems to be not necessary in this case, the port is always there.

File: labgrid/driver/power/netio_json.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        try:
+            assert 0 <= int(action) <= 4
+            return action
+        except AssertionError as ae:
+            raise Exception(f"{ae}\n Action \"{action}\" seems not right, should be between -> '0'-'4'!")
+
+
+def power_set(host, port, index, action):
+    """
+    set netio-state at index(socketID)
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py]:
```suggestion
    Set netio-state at index(socketID)
```

File: labgrid/driver/power/netio_json.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        - 3: Short ON delay,
+        - 4: Toggle (invert the state)
+    """
+    netio = __NetioControl(host, port)
+    print(netio.set_state(netio.convert_socket_id(index), action))
+
+
+def power_get(host, port, index):
+    """
+    get power-state as json
```
Comment (ChHapp) [inline labgrid/driver/power/netio_json.py]:
```suggestion
    Get power-state as json
```

**Review by ChHapp — CHANGES_REQUESTED** [review]

File: labgrid/driver/power/netio_json.py (hunk 3 above)
Comment (ActionHOSchT) [reply labgrid/driver/power/netio_json.py:62]: image

Ja, ist mit getestet.

File: labgrid/driver/power/netio_json.py (hunk 1 above)
Comment (ActionHOSchT) [reply labgrid/driver/power/netio_json.py:15]: isort forciert diese jedoch und ist demnach von pengutronics gefordert.

**Review by EugenWiens — APPROVED** [review]

File: labgrid/driver/power/netio_json.py (hunk 1 above)
Comment (ChHapp) [reply labgrid/driver/power/netio_json.py:15]: :ok_hand:

File: labgrid/driver/power/netio_json.py (hunk 2 above)
Comment (ActionHOSchT) [reply labgrid/driver/power/netio_json.py]: done

File: labgrid/driver/power/netio_json.py (hunk 4 above)
Comment (ActionHOSchT) [reply labgrid/driver/power/netio_json.py]: done

File: labgrid/driver/power/netio_json.py (hunk 5 above)
Comment (ActionHOSchT) [reply labgrid/driver/power/netio_json.py]: done

File: labgrid/driver/power/netio_json.py (hunk 6 above)
Comment (ActionHOSchT) [reply labgrid/driver/power/netio_json.py]: done

File: labgrid/driver/power/netio_json.py (hunk 7 above)
Comment (ActionHOSchT) [reply labgrid/driver/power/netio_json.py]: done
image

File: doc/configuration.rst [hunk 10]
```diff
@@ -192,6 +192,23 @@ Currently available are:
 ``netio``
   Controls a NETIO 4-Port PDU via a simple HTTP API.
 
+``netio_json``
+  Controls a NETIO n-Port PDU via a simple HTTP-JSON API.
+  .. code-block:: yaml
```
Comment (ChHapp) [inline doc/configuration.rst]: please, check the YAML formatting - my preview seems not right

Comment (ChHapp) [inline doc/configuration.rst]: please, insert an empty line here, thx

Comment (ActionHOSchT) [reply doc/configuration.rst]: done
image

Comment (ActionHOSchT) [reply doc/configuration.rst]: done

File: doc/configuration.rst [hunk 11]
```diff
@@ -192,6 +192,24 @@ Currently available are:
 ``netio``
   Controls a NETIO 4-Port PDU via a simple HTTP API.
 
+``netio_json``
+  Controls a NETIO n-Port PDU via a simple HTTP-JSON API.
+
+  .. code-block:: yaml
+
+    NetworkPowerPort:
```
Comment (ChHapp) [inline doc/configuration.rst]: By a closer look, do we really need this section here? It seems to me, it is already listed by the NetworkPowerDriver above.

**Review by ChHapp — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**